File: android/feature/auth/src/main/java/com/mplayeraudio/feature/auth/yamusic/YandexMusicAuthCardViewModel.kt
package com.mplayeraudio.feature.auth.yamusic

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.mplayeraudio.core.domain.yandexauth.YandexAuthException
import com.mplayeraudio.core.domain.yandexauth.YandexAuthStatus
import com.mplayeraudio.core.domain.yandexauth.YandexAuthorizationRequest
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

data class YandexMusicAuthCardState(
    val isLoading: Boolean = false,
    val isAuthorized: Boolean = false,
    val authorizationRequest: YandexAuthorizationRequest? = null,
)

sealed interface YandexMusicAuthCardEffect {
    data class ShowToast(
        val message: String,
    ) : YandexMusicAuthCardEffect
}

class YandexMusicAuthCardViewModel(
    private val startYandexAuthorization: StartYandexAuthorizationUseCase,
    private val completeYandexAuthorization: CompleteYandexAuthorizationUseCase,
    observeYandexSession: ObserveYandexSessionUseCase,
    observeYandexAuthStatus: ObserveYandexAuthStatusUseCase,
) : ViewModel() {
    private val _state = MutableStateFlow(YandexMusicAuthCardState())
    val state: StateFlow<YandexMusicAuthCardState> = _state.asStateFlow()

    private val _effects = MutableSharedFlow<YandexMusicAuthCardEffect>()
    val effects: SharedFlow<YandexMusicAuthCardEffect> = _effects.asSharedFlow()

    init {
        observeYandexSession()
            .onEach { session ->
                _state.update { currentState ->
                    currentState.copy(isAuthorized = session != null)
                }
            }
            .launchIn(viewModelScope)

        observeYandexAuthStatus()
            .onEach(::handleAuthStatus)
            .launchIn(viewModelScope)
    }

    fun onLoginClicked() {
        viewModelScope.launch {
            _state.update { currentState -> currentState.copy(isLoading = true) }

            try {
                val request = startYandexAuthorization()
                _state.update { currentState ->
                    currentState.copy(
                        isLoading = false,
                        authorizationRequest = request,
                    )
                }
            } catch (exception: YandexAuthException) {
                _state.update { currentState -> currentState.copy(isLoading = false) }
                _effects.emit(YandexMusicAuthCardEffect.ShowToast(exception.toUserMessage()))
            }
        }
    }

    fun onAuthorizationCallbackReceived(callbackUri: String) {
        viewModelScope.launch {
            _state.update { currentState -> currentState.copy(isLoading = true) }
            try {
                completeYandexAuthorization(callbackUri)
            } catch (_: YandexAuthException) {
                // Ошибка уже попадёт в observeYandexAuthStatus и превратится в пользовательское сообщение.
            }
        }
    }

    fun onAuthorizationCancelled() {
        _state.update { currentState ->
            currentState.copy(
                isLoading = false,
                authorizationRequest = null,
            )
        }
    }

    fun onAuthorizationLaunchFailed() {
        _state.update { currentState ->
            currentState.copy(
                isLoading = false,
                authorizationRequest = null,
            )
        }
        viewModelScope.launch {
            _effects.emit(
                YandexMusicAuthCardEffect.ShowToast(
                    "Не удалось запустить встроенную авторизацию Яндекса.",
                ),
            )
        }
    }

    internal fun dispose() {
        viewModelScope.cancel()
    }

    private fun handleAuthStatus(status: YandexAuthStatus) {
        when (status) {
            YandexAuthStatus.Authorizing -> {
                _state.update { currentState ->
                    currentState.copy(isLoading = currentState.authorizationRequest == null)
                }
            }

            is YandexAuthStatus.Authorized -> {
                _state.update { currentState ->
                    currentState.copy(
                        isLoading = false,
                        isAuthorized = true,
                        authorizationRequest = null,
                    )
                }
            }

            is YandexAuthStatus.Failed -> {
                _state.update { currentState ->
                    currentState.copy(
                        isLoading = false,
                        authorizationRequest = null,
                    )
                }
                viewModelScope.launch {
                    _effects.emit(
                        YandexMusicAuthCardEffect.ShowToast(
                            (status.error as? Throwable)?.toUserMessage()
                                ?: "Не удалось авторизоваться через Яндекс.",
                        ),
                    )
                }
            }

            YandexAuthStatus.Unauthorized -> {
                _state.update { currentState ->
                    currentState.copy(
                        isLoading = false,
                        authorizationRequest = null,
                    )
                }
            }
        }
    }

    private fun Throwable.toUserMessage(): String {
        return when (this) {
            YandexAuthException.MissingConfiguration -> {
                "Не настроен Yandex OAuth. Проверьте client_id, client_secret и redirect URI."
            }

            YandexAuthException.InvalidCallbackState -> {
                "Сессия авторизации устарела. Попробуйте войти ещё раз."
            }

            YandexAuthException.MissingAuthorizationCode -> {
                "Яндекс не вернул код авторизации."
            }

            is YandexAuthException.AccessDenied -> {
                "Доступ к аккаунту Яндекса не был подтверждён."
            }

            else -> {
                message ?: "Не удалось авторизоваться через Яндекс."
            }
        }
    }
}

File: ios/Feature/Auth/Sources/AuthFeature/Presentation/YandexMusicAuthCardViewModel.swift
import CoreDomain
import Foundation
import Observation

public struct YandexMusicAuthCardState: Equatable, Sendable {
    public var isLoading: Bool
    public var isAuthorized: Bool

    public init(
        isLoading: Bool = false,
        isAuthorized: Bool = false
    ) {
        self.isLoading = isLoading
        self.isAuthorized = isAuthorized
    }
}

public enum YandexMusicAuthCardEffect: Equatable, Sendable {
    case presentAuthWebView(YandexAuthorizationRequest)
    case showAlert(String)
}

@Observable
@MainActor
public final class YandexMusicAuthCardViewModel {
    public private(set) var state = YandexMusicAuthCardState()

    private let startYandexAuthorization: StartYandexAuthorizationUseCase
    private let completeYandexAuthorization: CompleteYandexAuthorizationUseCase
    private let cancelYandexAuthorization: CancelYandexAuthorizationUseCase
    private let observeYandexSession: ObserveYandexSessionUseCase
    private let observeYandexAuthStatus: ObserveYandexAuthStatusUseCase
    private let effectEmitter = AsyncEventEmitter<YandexMusicAuthCardEffect>()
    private var observationTask: Task<Void, Never>?
    private var isAwaitingAuthorizationCallback = false

    public init(
        startYandexAuthorization: StartYandexAuthorizationUseCase,
        completeYandexAuthorization: CompleteYandexAuthorizationUseCase,
        cancelYandexAuthorization: CancelYandexAuthorizationUseCase,
        observeYandexSession: ObserveYandexSessionUseCase,
        observeYandexAuthStatus: ObserveYandexAuthStatusUseCase
    ) {
        self.startYandexAuthorization = startYandexAuthorization
        self.completeYandexAuthorization = completeYandexAuthorization
        self.cancelYandexAuthorization = cancelYandexAuthorization
        self.observeYandexSession = observeYandexSession
        self.observeYandexAuthStatus = observeYandexAuthStatus
    }

    public func start() {
        guard observationTask == nil else { return }

        observationTask = Task { [weak self] in
            guard let self else { return }

            await withTaskGroup(of: Void.self) { group in
                group.addTask { [weak self] in
                    guard let self else { return }
                    for await session in observeYandexSession() {
                        if Task.isCancelled {
                            break
                        }
                        await self.handleSession(session)
                    }
                }

                group.addTask { [weak self] in
                    guard let self else { return }
                    for await status in observeYandexAuthStatus() {
                        if Task.isCancelled {
                            break
                        }
                        await self.handleAuthStatus(status)
                    }
                }
            }
        }
    }

    public func effectStream() -> AsyncStream<YandexMusicAuthCardEffect> {
        effectEmitter.stream()
    }

    public func onLoginTapped() async {
        state.isLoading = true

        do {
            let request = try await startYandexAuthorization()
            isAwaitingAuthorizationCallback = true
            effectEmitter.yield(.presentAuthWebView(request))
        } catch let error as YandexAuthException {
            state.isLoading = false
            isAwaitingAuthorizationCallback = false
            effectEmitter.yield(.showAlert(error.userMessage))
        } catch {
            state.isLoading = false
            isAwaitingAuthorizationCallback = false
            effectEmitter.yield(.showAlert(error.localizedDescription))
        }
    }

    public func onAuthWebViewLaunchFailed() async {
        state.isLoading = false
        isAwaitingAuthorizationCallback = false
        await cancelYandexAuthorization()
        effectEmitter.yield(.showAlert("Не удалось открыть страницу авторизации Яндекса."))
    }

    public func onAuthorizationCallback(_ callbackURL: URL) async {
        guard isAwaitingAuthorizationCallback else {
            return
        }

        isAwaitingAuthorizationCallback = false
        _ = try? await completeYandexAuthorization(callbackURL: callbackURL)
    }

    public func onAuthorizationDismissedWithoutCallback() async {
        guard isAwaitingAuthorizationCallback else {
            return
        }

        isAwaitingAuthorizationCallback = false
        await cancelYandexAuthorization()
    }

    private func handleSession(_ session: YandexAuthSession?) {
        state.isAuthorized = session != nil
    }

    private func handleAuthStatus(_ status: YandexAuthStatus) {
        switch status {
        case .unauthorized:
            state.isLoading = false
            isAwaitingAuthorizationCallback = false

        case .authorizing:
            state.isLoading = true

        case .authorized:
            state.isLoading = false
            state.isAuthorized = true
            isAwaitingAuthorizationCallback = false

        case let .failed(error):
            state.isLoading = false
            isAwaitingAuthorizationCallback = false
            effectEmitter.yield(.showAlert(error.userMessage))
        }
    }
}

private extension YandexAuthException {
    var userMessage: String {
        switch self {
        case .missingConfiguration:
            return "Не настроен Yandex OAuth. Проверьте client_id."

        case .invalidCallbackState:
            return "Сессия авторизации устарела. Попробуйте войти ещё раз."

        case .accessDenied:
            return "Доступ к аккаунту Яндекса не был подтверждён."

        default:
            return errorDescription ?? "Не удалось авторизоваться через Яндекс."
        }
    }
}
